### DISCUSSION TITLE
http requests to protected next.js api routes

### DISCUSSION THREAD

**Discussion by KyrillGobber** [supabase/supabase#13791] [Questions] [answered]
Hi there! I'm quite new to supabase and I'm trying to test my api routes with curls, but it seems like, even though req has the correct token in the headers, the supabase client will not.

[I followed this to setup a protected api route.](https://supabase.com/docs/guides/auth/auth-helpers/nextjs#protecting-api-routes)

Here's my code
```
const handler: NextApiHandler = async (req: NextApiRequest, res: NextApiResponse) => {
    const supabaseServerClient = createServerSupabaseClient({
        req,
        res,
    });
    const {
        data: { session },
    } = await supabaseServerClient.auth.getSession();

    if (!session)
        return res.status(401).json({
            error: 'not_authenticated',
            description:
                'The user does not have an active session or is not authenticated',
        });

    switch (req.method) {
        case 'GET':
            if (session?.user.id) return getData(res, session?.user.id);
            break;
        case 'POST':
            if (session?.user.id) {
                const dataToPaste = JSON.parse(req.body).title;
                return addNewData(res, title, session.user.id);
            }
            break;
        default:
            return res.status(404).json({ error: 'API method not found' });
    }
}
```

I tried both, with the authorization header alone, and with the apikey:
```
  curl 'http://localhost:3000/api/route' \
      -H "Authorization: Bearer [JWT]"

  curl 'http://localhost:3000/api/route' \
      -H "apikey: [apikey]" \
      -H "Authorization: Bearer [JWT]"
```

I am getting the token from a logged in user on the browser. If I curl supabase directly, it works. So it seems to be a problem of createServerSupabaseClient.

What am I doing wrong here?

**Comment by KyrillGobber** [comment, marked as answer]
After looking into this again I finally  found a solution: The supabase client does not check the authorization header, it checks the cookies. That means yo gotta curl like this:

```
curl 'http://localhost:3000/api/route' \
            -H "cookie: supabase-auth-token=[The supabase-auth-token you'll find in your cookie storage]"
```

This also applies if you're using rest.nvim or postman or whatever..

For example, with rest.nvim I'll use:

```
GET http://localhost:3000/api/route
cookie: supabase-auth-token=[The supabase-auth-token you'll find in your cookie storage]
```

The API Key seems to be irrelevant.

All clear, case closed.

**Comment by benank** [comment]
If anyone comes looking for how to access protected routes from any client, here's the magic sauce:

In SvelteKit backend, add this to `hooks.server.ts` (likely similar for other backends):
```
        event.locals.supabase.auth.setSession({
            access_token: accessToken,
            refresh_token: refreshToken
        })
```

You can provide accessToken and refreshToken through the headers in the request. Using this method, you can authenticate using any HTTP client, such as Postman using headers. The response will also set the supabase auth cookie for future requests.

Provide your access token and refresh token from wherever you signed in from.
